### ISSUE TITLE
Match concrete variable name

### ISSUE THREAD

**Issue by sshine** [ndmitchell/hlint#611]
I'd like to write HLint hints for specific programming exercises.

One case is naming a variable `y` when it should preferrably be named `year`.

    hint:
      lhs: 'y'
      note: You might as well call the variable 'year'.
      rhs: year

Unfortunately this `lhs` pattern becomes a catch-all expression rather than the literal variable name `y`. Is there any way to match the concrete variable name `y`, and if not, would it be interesting to extend HLint with this capability?

**Comment by ndmitchell** [comment]
Interesting! Hadn't thought about that before. My current thoughts are to use the `side` conditions so you could write:

```
hint:
  lhs: 'y'
  note: You might as well call the variable 'year'.
  rhs: year
  side: is "y" y
```

That seems about the cleanest I can come up with. Does that seem reasonable? If so, we can add the `is "y"` bit to `side`.

**Comment by sshine** [comment]
If I understand what you're saying correctly, `side:` conditions are what's described in [`Hint.Match`](https://github.com/ndmitchell/hlint/blob/master/src/Hint/Match.hs), and you're proposing that we add another side condition, `is "foo" bar` that fixes `lhs: 'foo'` from being a variable pattern to being a concrete variable named `bar`?

Before I say yes, maybe I should assess whether HLint is really suited to give this type of hints; here are some other hints that I find hard to encode:

- `rem x y` → ``x `rem` y``. The following triggers even though I *have* written ``x `rem` y``:

      - hint:
          lhs: rem x y
          note: You can write 'rem' as an infix operator.
          rhs: x `rem` y

- ``(year `rem` 4) == 0`` → ``year `rem` 4 == 0``. I'm not sure why HLint doesn't already catch this.

  Backticks have precedence level 7, `==` has 4.

Then there are hints that altogether don't lend themselves towards HLint's simplicity, such as "you may want to use guards for this function", "you may want to replace calling `f` directly with a helper function here", "since the other function is camelCased, this one ought to be, too", and "the order of guards is not optimal".

I don't know if it's easier to extend HLint or write something based directly on `haskell-src-exts`.

**Comment by ndmitchell** [comment]
You certainly could write your own tool based on haskell-src-exts, but I've just added enough to HLint to catch everything you want given the rules:

```yaml
- hint: {lhs: "y", rhs: year, side: subst y == y}
- hint: {lhs: rem x y, rhs: x `rem` y, side: original == subst (x `rem` y)}
# alternatively - hint: {lhs: rem x y, rhs: x `rem` y, side: not (isInfixApp original)}
- hint: {lhs: (a `rem` b) == c, rhs: a `rem` b == c, side: subst original == subst ((a `rem` b) == c)}
```

In particular I added `==` to the `side` conditions, and an explicit `subst`, so you can write whatever test you want.

Note that HLint never assumes fixity when removing brackets because often people put brackets around different fixity stuff for clarity and it doesn't actually know the fixity of everything (sometimes it guesses).

If this does work for you, I'd love to see a blog post or other write up of how you configured HLint to help students. I think it could be the kind of thing that is useful for lots of people.

I haven't made a release yet - but can do so if that helps testing/use.

**Comment by sshine** [comment]
Thanks a lot for your effort! I'll look at this tomorrow.

**Comment by sshine** [comment]
The famous "tomorrow". I still have this project on my radar. Since the subject of the issue is "Match concrete variable names" and you showed how I can do this, I'm closing the issue. I will get back to you with a story of how using `hlint` for exercise-specific hints once I get back to this. Thanks a lot for your help so far!

**Comment by masaeedu** [comment]
Perhaps it makes sense to leave this open instead, so that anyone who wants to pick it up can do so.

**Comment by ndmitchell** [comment]
@masaeedu The questioner wanted to match concrete variable names. HLint added features to do that. I don't think there's anything more for HLint to do (you could argue some docs wouldn't go amiss...)
